### ISSUE TITLE
Exception("DummyFuture does not support blocking for results")

### ISSUE THREAD

**Issue by TheBigBear** [saltstack/salt#26884]
On a 2015.8 master running against a 2015.8 minion (full salt --versions output follows at bottom) I get an error raised if it tries to copy a file from master to the minion cache.

At first I forgot to have the installer file `windirstat1_1_2_setup.exe` in the master's `win/repo/windirstat` directory so it gave me the following error, which **_is correct_** and expected, since the installer file is missing on the master.

```
salt 'om-uk-*w*' pkg.install windirstat
om-uk-w10-1-64:
    ----------
    windirstat:
        ----------
        unable to cache:
            salt://win/repo/windirstat/windirstat1_1_2_setup.exe
om-uk-w81-1-64:
    ----------
    windirstat:
        ----------
        unable to cache:
            salt://win/repo/windirstat/windirstat1_1_2_setup.exe
om-uk-w7-1-64:
    ----------
    windirstat:
        ----------
        unable to cache:
            salt://win/repo/windirstat/windirstat1_1_2_setup.exe
```

But then I put the installer file `windirstat1_1_2_setup.exe` on the master and now I get:

```
salt 'om-uk-*w*' pkg.install windirstat
om-uk-w7-1-64:
    The minion function caused an exception: Traceback (most recent call last):
      File "c:\salt-dev\salt\salt\minion.py", line 1004, in _thread_return
        return_data = func(*args, **kwargs)
      File "c:\salt-dev\salt\salt\modules\win_pkg.py", line 602, in install
        cached_pkg = __salt__['cp.cache_file'](installer, saltenv)
      File "c:\salt-dev\salt\salt\modules\cp.py", line 365, in cache_file
        result = __context__['cp.fileclient'].cache_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 155, in cache_file
        return self.get_url(path, '', True, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 540, in get_url
        return self.get_file(url, dest, makedirs, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 996, in get_file
        hash_server = self.hash_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 1236, in hash_file
        return self.channel.send(load)
      File "c:\salt-dev\salt\salt\utils\async.py", line 73, in wrap
        ret = self._block_future(ret)
      File "c:\salt-dev\salt\salt\utils\async.py", line 83, in _block_future
        return future.result()
      File "c:\python27\lib\site-packages\tornado\concurrent.py", line 216, in result
        self._check_done()
      File "c:\python27\lib\site-packages\tornado\concurrent.py", line 294, in _check_done
        raise Exception("DummyFuture does not support blocking for results")
    Exception: DummyFuture does not support blocking for results
```

Master `salt --versions`: git '2015.8' branch on 2 Sept at commit https://github.com/saltstack/salt/commit/f68ba2853a5fbf41f70744410576dbff33011302

```
salt --versions
Salt Version:
           Salt: 2015.8.0

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: 0.21.1
           Mako: 0.8.1
         PyYAML: 3.11
          PyZMQ: 14.7.0
         Python: 2.7.5 (default, Jun 24 2015, 00:41:19)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: 3.2.2
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: Not Installed
          ioflo: 1.3.9
        libnacl: Not Installed
   msgpack-pure: 0.1.3
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: 0.21.4
   python-gnupg: Not Installed
          smmap: 0.8.1
        timelib: 0.2.4

System Versions:
           dist: centos 7.1.1503 Core
        machine: x86_64
        release: 3.10.0-229.11.1.el7.x86_64
         system: CentOS Linux 7.1.1503 Core
```

and the 3 windows minion are all:

```
C:\Salt-Dev>salt-call --versions
Salt Version:
           Salt: 2015.8.0rc3-181-ga21b3f4

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
         Python: 2.7.8 (default, Jun 30 2014, 16:08:48) [MSC v.1500 64 bit (AMD6
4)]
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.2
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist:
        machine: AMD64
        release: 7
         system: 7 6.1.7601 SP1 Multiprocessor Free
```

**Comment by TheBigBear** [comment]
OK, a quick update. I revved up the versions on both master and minion to ```2015.8.0rc3-1154-g8cdf75c (Beryllium)```, and I still get issues, but they have shifted ever so slightly.

I am now using a new winrepo-ng package (```curl```) that also requires it's installer to be on the master 'salt://win/repo' for the minion to be able to install it.

Here are my steps - and the results.

1 started off again with NOT having the installer on the masters repo.
```
time salt 'om-uk-w*' pkg.install curl
om-uk-w7-1-64:
    The minion function caused an exception: Traceback (most recent call last):
      File "c:\salt-dev\salt\salt\minion.py", line 1004, in _thread_return
        return_data = func(*args, **kwargs)
      File "c:\salt-dev\salt\salt\modules\win_pkg.py", line 602, in install
        cached_pkg = __salt__['cp.cache_file'](installer, saltenv)
      File "c:\salt-dev\salt\salt\modules\cp.py", line 365, in cache_file
        result = __context__['cp.fileclient'].cache_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 155, in cache_file
        return self.get_url(path, '', True, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 540, in get_url
        return self.get_file(url, dest, makedirs, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 996, in get_file
        hash_server = self.hash_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 1236, in hash_file
        return self.channel.send(load)
      File "c:\salt-dev\salt\salt\utils\async.py", line 73, in wrap
        ret = self._block_future(ret)
      File "c:\salt-dev\salt\salt\utils\async.py", line 83, in _block_future
        return future.result()
      File "c:\python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "c:\python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 220, in send
        ret = yield self._crypted_transfer(load, tries=tries, timeout=timeout)
      File "c:\python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "c:\python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "c:\python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 193, in _crypted_transfer
        ret = yield _do_transfer()
      File "c:\python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "c:\python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "c:\python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 179, in _do_transfer
        timeout=timeout,
      File "c:\python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "c:\python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
    SaltReqTimeoutError: Message timed out
om-uk-w81-1-64:
    The minion function caused an exception: Traceback (most recent call last):
      File "c:\salt-dev\salt\salt\minion.py", line 1004, in _thread_return
        return_data = func(*args, **kwargs)
      File "c:\salt-dev\salt\salt\modules\win_pkg.py", line 602, in install
        cached_pkg = __salt__['cp.cache_file'](installer, saltenv)
      File "c:\salt-dev\salt\salt\modules\cp.py", line 365, in cache_file
        result = __context__['cp.fileclient'].cache_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 155, in cache_file
        return self.get_url(path, '', True, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 540, in get_url
        return self.get_file(url, dest, makedirs, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 996, in get_file
        hash_server = self.hash_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 1236, in hash_file
        return self.channel.send(load)
      File "c:\salt-dev\salt\salt\utils\async.py", line 73, in wrap
        ret = self._block_future(ret)
      File "c:\salt-dev\salt\salt\utils\async.py", line 83, in _block_future
        return future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 220, in send
        ret = yield self._crypted_transfer(load, tries=tries, timeout=timeout)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 193, in _crypted_transfer
        ret = yield _do_transfer()
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 179, in _do_transfer
        timeout=timeout,
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
    SaltReqTimeoutError: Message timed out
om-uk-w10-1-64:
    The minion function caused an exception: Traceback (most recent call last):
      File "c:\salt-dev\salt\salt\minion.py", line 1004, in _thread_return
        return_data = func(*args, **kwargs)
      File "c:\salt-dev\salt\salt\modules\win_pkg.py", line 602, in install
        cached_pkg = __salt__['cp.cache_file'](installer, saltenv)
      File "c:\salt-dev\salt\salt\modules\cp.py", line 365, in cache_file
        result = __context__['cp.fileclient'].cache_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 155, in cache_file
        return self.get_url(path, '', True, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 540, in get_url
        return self.get_file(url, dest, makedirs, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 996, in get_file
        hash_server = self.hash_file(path, saltenv)
      File "c:\salt-dev\salt\salt\fileclient.py", line 1236, in hash_file
        return self.channel.send(load)
      File "c:\salt-dev\salt\salt\utils\async.py", line 73, in wrap
        ret = self._block_future(ret)
      File "c:\salt-dev\salt\salt\utils\async.py", line 83, in _block_future
        return future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 220, in send
        ret = yield self._crypted_transfer(load, tries=tries, timeout=timeout)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 193, in _crypted_transfer
        ret = yield _do_transfer()
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 876, in run
        yielded = self.gen.throw(*exc_info)
      File "c:\salt-dev\salt\salt\transport\zeromq.py", line 179, in _do_transfer
        timeout=timeout,
      File "C:\Python27\lib\site-packages\tornado\gen.py", line 870, in run
        value = future.result()
      File "C:\Python27\lib\site-packages\tornado\concurrent.py", line 215, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
    SaltReqTimeoutError: Message timed out

real	1m2.224s
user	0m2.186s
sys	0m0.376s
```

BUT I no longer see the error about "Exception("DummyFuture does not support blocking for results")" , but it still looks 'un-clean'?

Anyway, I then preceeded to re-run this exact same cmd still without ther curl installer present on the master and it now runs through without the above errors. I tried severeal times and the errors only ever showed up the very first time. Really strange - eh?

```
time salt 'om-uk-*' pkg.install curl
om-uk-w81-1-64:
    ----------
    curl:
        ----------
        unable to cache:
            salt://win/repo/curl/curl-7.43.0-win64.msi
om-uk-w7-1-64:
    ----------
    curl:
        ----------
        unable to cache:
            salt://win/repo/curl/curl-7.43.0-win64.msi
om-uk-w10-1-64:
    ----------
    curl:
        ----------
        unable to cache:
            salt://win/repo/curl/curl-7.43.0-win64.msi

real	0m23.210s
user	0m1.694s
sys	0m0.207s
```
and I repeated this severeal times in a row and it alwsy came back cleanly with these mesages that were 100% correct, it can't cache a file that is not there.

And if I now re-run the install with the installer in place on the master.
```
om-uk-w7-1-64:
    ----------
    curl:
        ----------
        new:
            7.43.0
        old:
om-uk-w81-1-64:
    ----------
    curl:
        ----------
        new:
            7.43.0
        old:
om-uk-w10-1-64:
    ----------
    curl:
        ----------
        new:
            7.43.0
        old:
```

So in summary, there is a remaining issue where __IF__ and __WHEN__ the installer is not present on the installer, but seemingly only on the very first installation attempt.

**Comment by redbaron4** [comment]
I have been seeing this error ever since I upgraded to 2018.3.0. I didn't have this problem in earlier salt versions. Maybe the change in tornado version created an incompatibility. I am on Python-2.6

**Comment by rallytime** [comment]
@redbaron4 Python 2.6 is no lot supported in the 2018.3.x releases. You'll need to use 2.7 or >= 3.4.

**Comment by redbaron4** [comment]
@rallytime Ah sorry..I forgot I had updated all infra to 2.7 before moving to 2017.7 itself. So I am on python - 2.7 (RHEL6)

**Comment by rallytime** [comment]
@redbaron4 Ah, OK. Thanks!
